database/engine/rrdengineapi.c

```diff
@@ -75,97 +103,128 @@ void rrdeng_convert_legacy_uuid_to_multihost(char machine_guid[GUID_LEN + 1], uu
-    PValue = JudyHSGet(pg_cache->metrics_index.JudyHS_array, &legacy_uuid, sizeof(uuid_t));
+    Pvoid_t *PValue = JudyHSGet(pg_cache->metrics_index.JudyHS_array, uuid, sizeof(uuid_t));
     if (likely(NULL != PValue)) {
         page_index = *PValue;
+        page_index->refcount++;
+
+        if(page_index->ctx && page_index->ctx != ctx)
+            fatal("DBENGINE: page_index has a different ctx.");
+
+        if(!page_index->ctx)
+            page_index->ctx = ctx;
+
+        if(pa) {
+            if(page_index->alignment && page_index->alignment != pa)
+                fatal("DBENGINE: page_index has a different alignment.");
+
+            if(!page_index->alignment) {
+                page_index->alignment = pa;
+                pa->refcount++;
+            }
+        }
     }
     uv_rwlock_rdunlock(&pg_cache->metrics_index.lock);
-    if (is_multihost_child || NULL == PValue) {
-        /* First time we see the legacy UUID or metric belongs to child host in multi-host DB.
-         * Drop legacy support, normal path */
-
-        uv_rwlock_rdlock(&pg_cache->metrics_index.lock);
-        PValue = JudyHSGet(pg_cache->metrics_index.JudyHS_array, &rd->metric_uuid, sizeof(uuid_t));
-        if (likely(NULL != PValue)) {
-            page_index = *PValue;
-        }
-        uv_rwlock_rdunlock(&pg_cache->metrics_index.lock);
-        if (NULL == PValue) {
-            uv_rwlock_wrlock(&pg_cache->metrics_index.lock);
-            PValue = JudyHSIns(&pg_cache->metrics_index.JudyHS_array, &rd->metric_uuid, sizeof(uuid_t), PJE0);
-            fatal_assert(NULL == *PValue); /* TODO: figure out concurrency model */
-            *PValue = page_index = create_page_index(&rd->metric_uuid);
-            page_index->prev = pg_cache->metrics_index.last_page_index;
-            pg_cache->metrics_index.last_page_index = page_index;
-            uv_rwlock_wrunlock(&pg_cache->metrics_index.lock);
-        }
-    } else {
-        /* There are legacy UUIDs in the database, implement backward compatibility */
 
-        rrdeng_convert_legacy_uuid_to_multihost(rd->rrdset->rrdhost->machine_guid, &legacy_uuid,
-                                                &multihost_legacy_uuid);
+    return (STORAGE_METRIC_HANDLE *)page_index;
+}
+
+STORAGE_METRIC_HANDLE *rrdeng_metric_create(STORAGE_INSTANCE *db_instance, uuid_t *uuid, STORAGE_METRICS_GROUP *smg) {
+    struct rrdengine_instance *ctx = (struct rrdengine_instance *)db_instance;
+    struct pg_alignment *pa = (struct pg_alignment *)smg;
+    struct pg_cache_page_index *page_index;
+    struct page_cache *pg_cache = &ctx->pg_cache;
 
-        int need_to_store = uuid_compare(rd->metric_uuid, multihost_legacy_uuid);
+    uv_rwlock_wrlock(&pg_cache->metrics_index.lock);
+    Pvoid_t *PValue = JudyHSIns(&pg_cache->metrics_index.JudyHS_array, uuid, sizeof(uuid_t), PJE0);
+    fatal_assert(NULL == *PValue); /* TODO: figure out concurrency model */
+    *PValue = page_index = create_page_index(uuid);
```

Comment:
We also have this:

```sh
176         pg_cache->metrics_index.last_page_index = page_index;
177         page_index->alignment = pa;
178         if(pa)
179             pa->refcount++;
180         uv_rwlock_wrunlock(&pg_cache->metrics_index.lock);
181     
>>>     CID 358387:  Resource leaks  (RESOURCE_LEAK)
>>>     Variable "PValue" going out of scope leaks the storage it points to.
182         return (STORAGE_METRIC_HANDLE *)page_index;
183     }
184     
185     STORAGE_METRIC_HANDLE *rrdeng_metric_get_or_create(RRDDIM *rd, STORAGE_INSTANCE *db_instance, STORAGE_METRICS_GROUP *smg) {
186         STORAGE_METRIC_HANDLE *db_metric_handle;
187     
```

I think we already talked about this in previous PR, but I am bringing these leaks that Judy adds to bring a complete report.

Reply:
this is how judy works